streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
+                                                             final Map<KeyValue<String, String>, Set<ProcessId>> tagEntryToClients,
+                                                             final Map<String, Set<String>> tagKeyToValues,
+                                                             final Map<KeyValue<String, String>, Set<ProcessId>> tagEntryToUsedClients,
+                                                             final Function<KafkaStreamsState, Map<String, String>> clientTagGetter) {
+        final Map<String, String> usedClientTags = clientTagGetter.apply(usedClient);
+
+        for (final Map.Entry<String, String> usedClientTagEntry : usedClientTags.entrySet()) {
+            final String tagKey = usedClientTagEntry.getKey();
+
+            if (!rackAwareAssignmentTags.contains(tagKey)) {
+                LOG.warn("Client tag with key [{}] will be ignored when computing rack aware standby " +
+                         "task assignment because it is not part of the configured rack awareness [{}].",
+                    tagKey, rackAwareAssignmentTags);
+                continue;
+            }
+
+            final Set<String> allTagValues = tagKeyToValues.get(tagKey);
+
+            if (allTagValues.size() <= countOfUsedClients) {
+                allTagValues.forEach(tagValue -> tagEntryToUsedClients.remove(new KeyValue<>(tagKey, tagValue)));
+            } else {
+                final String tagValue = usedClientTagEntry.getValue();
+                final KeyValue<String, String> tagEntry = new KeyValue<>(tagKey, tagValue);
+                final Set<ProcessId> clientsOnUsedTagValue = tagEntryToClients.get(tagEntry);
+                tagEntryToUsedClients.put(tagEntry, clientsOnUsedTagValue);
+            }
+        }
+    }
+
+    private static Function<KafkaStreamsState, Map<String, String>> createClientTagGetter(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+        final boolean canPerformRackAwareOptimization = canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY);
+
+        if (hasRackAwareAssignmentTags || !canPerformRackAwareOptimization) {
+            return KafkaStreamsState::clientTags;
+        } else {
+            return state -> mkMap(mkEntry("rack", state.rackId().get()));
+        }
+    }
+
+    private static List<String> getRackAwareAssignmentTags(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+
+        if (hasRackAwareAssignmentTags) {
+            return applicationState.assignmentConfigs().rackAwareAssignmentTags();
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return Collections.singletonList("rack");
```

Comment:
Similar to [this comment](https://github.com/apache/kafka/pull/16129/files#r1619754391), I don't think the middle case exists in non-testing code, and we shouldn't have special cases for tests in production code.

Also, once you make the simplification from that comment, it should only be possible to reach this if there are rack aware tags. So we should be able to just inline the `applicationState.assignmentConfigs().rackAwareAssignmentTags()` and get rid of this method altogether